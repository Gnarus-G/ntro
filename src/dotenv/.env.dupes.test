# comment
# comment
NAME="value"

# @type 'a' 
NAME2=b
KEY = "value" # asdfa


# as;dfkj
# as;dfkj
# @type string
keys= 'city'


# @type number
keys2 ='city'
