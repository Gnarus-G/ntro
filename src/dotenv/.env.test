# comment
# comment
# IGNORED_KEY='whatever'
NAME="value"

# @type 'a' | 'b'
NAME2=value
KEY = "value" # asdfa

# @type 'qa' | 'development' | 'production'
NEXT_PUBLIC_ENVIRONMENT = 'development'

# as;dfkj
# as;dfkj
keys= 'city'


# @type string
keys2 ='city'
