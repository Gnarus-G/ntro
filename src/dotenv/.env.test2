# comment
# comment
NAME5=value
# @type number
KEY_Value=123 # pair

# @type 'city' | 'townhall'
keys2Da='city'

# @type 'val' | 'value'
NAME = "value"

NEXT_PUBLIC_ENVIRONMENT = 'qa'

