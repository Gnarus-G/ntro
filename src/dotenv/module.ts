import z, { ZodTypeAny } from "zod";

const clientEnvSchemas = {
  dummy: z.string(),
};
const serverEnvSchemas = {
  dummy: z.string(),
};

const processEnv = {};

/* --- MAIN IMPLEMENTATION BELOW --- */

type ClientEnv = z.infer<z.ZodObject<typeof clientEnvSchemas>>;

export const clientEnv: ClientEnv = new Proxy({} as ClientEnv, {
  get(_, prop: string) {
    return lookupEnv(prop, clientEnvSchemas, () => {
      throw new Error(
        `${prop} is not defined for client side environment variables.`
      );
    });
  },
});

type Env = z.infer<z.ZodObject<typeof serverEnvSchemas>>;

export const env: Env = new Proxy({} as Env, {
  get(_, prop: string) {
    if (prop.startsWith("NEXT_PUBLIC_")) {
      return Reflect.get(clientEnv, prop);
    }
    return lookupEnv(prop, serverEnvSchemas, () => {
      throw new Error(
        `${prop} is not defined for server side environment variables.`
      );
    });
  },
});

const cache: Record<string, unknown> = {};

function lookupEnv<T extends Record<string, ZodTypeAny>>(
  prop: string,
  parsers: T,
  onNotFound: () => never
) {
  if (prop in cache) {
    return cache[prop];
  }

  try {
    if (prop in parsers) {
      const parsed = parsers[prop as keyof typeof parsers]?.parse(
        processEnv[prop as keyof typeof processEnv],
        { path: [prop] }
      );

      cache[prop] = parsed;

      return parsed;
    }
    onNotFound();
  } catch (e) {
    throw new BadEnvError(`failed to read ${prop} from proccess.env`, e);
  }
}

class BadEnvError extends Error {
  constructor(public message: string, public cause: unknown) {
    super(message);
    if (cause instanceof Error) {
      this.message = [message, cause].join("\n â†³ ");
    }
  }
}
