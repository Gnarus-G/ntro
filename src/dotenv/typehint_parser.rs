use std::fmt::Display;

#[derive(Debug, PartialEq, Eq, Clone)]
pub enum TypeHint {
    String,
    Number,
    Boolean,
    Union(Box<[Box<str>]>),
}

impl Display for TypeHint {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            TypeHint::Union(values) => values
                .iter()
                .map(|a| a.as_ref())
                .collect::<Vec<_>>()
                .join(" | "),
            tk => format!("{tk:?}").to_lowercase(),
        };

        f.write_str(&s)
    }
}

pub trait ParseTyeHint {
    fn into_type_hint(self) -> Option<TypeHint>;
}

impl ParseTyeHint for &str {
    fn into_type_hint(self) -> Option<TypeHint> {
        Parser::new(self).parse().ok()
    }
}

#[derive(PartialEq, Debug, Clone, Copy)]
enum TokenKind {
    Keyword,
    Pound,
    StringType,
    NumberType,
    BooleanType,
    StringLiteral,
    Pipe,
    Eof,
    Illegal,
}

#[derive(Debug, Clone, Copy)]
struct Token<'source> {
    kind: TokenKind,
    text: &'source str,
}

struct Lexer<'source> {
    source: &'source str,
    position: usize,
}

impl<'source> Lexer<'source> {
    pub fn new(source: &'source str) -> Self {
        Self {
            position: 0,
            source,
        }
    }

    fn char(&self) -> Option<&'source u8> {
        self.source.as_bytes().get(self.position)
    }

    fn char_skipping_whitespace(&mut self) -> Option<&'source u8> {
        while self
            .char()
            .map(|&c| c.is_ascii_whitespace())
            .unwrap_or(false)
        {
            self.step();
        }

        return self.char();
    }

    fn step(&mut self) {
        self.position += 1;
    }

    pub fn next_token(&mut self) -> Token<'source> {
        let Some(ch) = self.char_skipping_whitespace() else {
            return Token{
                kind: TokenKind::Eof,
                text: ""
            }
        };

        let token = match ch {
            b'@' => self.lex_keyword("@type"),
            b'\'' => self.lex_string_literal(),
            b'|' => Token {
                kind: TokenKind::Pipe,
                text: "|",
            },
            c if c.is_ascii_alphabetic() => self.lex_type(),
            b'#' => Token {
                kind: TokenKind::Pound,
                text: "#",
            },
            _ => Token {
                kind: TokenKind::Illegal,
                text: &self.source[self.position..self.position + 1],
            },
        };

        self.step();

        token
    }

    fn lex_type(&mut self) -> Token<'source> {
        let start = self.position;

        while self
            .char()
            .map(|&c| c.is_ascii_lowercase())
            .unwrap_or(false)
        {
            self.step();
        }

        let s = &self.source[start..self.position];

        match s {
            "string" => Token {
                kind: TokenKind::StringType,
                text: s,
            },
            "number" => Token {
                kind: TokenKind::NumberType,
                text: s,
            },
            "boolean" => Token {
                kind: TokenKind::BooleanType,
                text: s,
            },
            _ => Token {
                kind: TokenKind::Illegal,
                text: s,
            },
        }
    }

    fn lex_keyword(&mut self, keyword: &str) -> Token<'source> {
        let start = self.position;

        self.step();

        while self
            .char()
            .map(|&c| c.is_ascii_lowercase())
            .unwrap_or(false)
        {
            self.step();
        }

        let s = &self.source[start..self.position];

        if s == keyword {
            return Token {
                kind: TokenKind::Keyword,
                text: s,
            };
        }

        return Token {
            kind: TokenKind::Illegal,
            text: s,
        };
    }

    fn lex_string_literal(&mut self) -> Token<'source> {
        let start = self.position;

        self.step();

        while self.char().map(|&c| c != b'\'').unwrap_or(false) {
            self.step();
        }

        let Some(b'\'') = self.char() else {
            let s = &self.source[start..self.position];
            return Token {
                kind: TokenKind::Illegal,
                text: s,
            };
        };

        self.step();

        let s = &self.source[start..self.position];

        return Token {
            kind: TokenKind::StringLiteral,
            text: s,
        };
    }
}

impl<'source> Iterator for Lexer<'source> {
    type Item = Token<'source>;

    fn next(&mut self) -> Option<Self::Item> {
        let token = self.next_token();

        if token.kind == TokenKind::Eof {
            return None;
        }

        return Some(token);
    }
}

use thiserror::Error;

#[derive(Debug, Error)]
enum ParseError<'source> {
    #[error("expected to find {expected:?} but found {found:?}")]
    ExpectedToken {
        expected: TokenKind,
        found: Token<'source>,
    },
    #[error("unexpected end of input")]
    UnexpectedEnd,
    #[error("unexpected token found: {found:?}")]
    IllegalToken { found: Token<'source> },
}

struct Parser<'source> {
    lexer: Lexer<'source>,
    token: Token<'source>,
    peeked: Option<Token<'source>>,
}

impl<'source> Parser<'source> {
    pub fn new(code: &'source str) -> Self {
        let mut lexer = Lexer::new(code);
        Self {
            peeked: None,
            token: lexer.next_token(),
            lexer,
        }
    }

    fn next_token(&mut self) -> &Token<'source> {
        self.token = match self.peeked.take() {
            Some(t) => t,
            None => self.lexer.next_token(),
        };

        &self.token
    }

    // fn peek_token(&mut self) -> &Token<'source> {
    //     self.peeked.get_or_insert_with(|| self.lexer.next_token())
    // }

    pub fn parse(&mut self) -> Result<TypeHint, ParseError<'source>> {
        if self.token.kind == TokenKind::Pound {
            self.next_token();
        }

        self.expect(TokenKind::Keyword)?;

        self.next_token();

        match self.token.kind {
            TokenKind::StringType => return Ok(TypeHint::String),
            TokenKind::NumberType => return Ok(TypeHint::Number),
            TokenKind::BooleanType => return Ok(TypeHint::Boolean),
            TokenKind::StringLiteral => {
                let mut union: Vec<Box<str>> = vec![self.token.text.into()];

                while self.next_token().kind == TokenKind::Pipe && self.token.kind != TokenKind::Eof
                {
                    // just ignore any bunch of consecutive pipes
                    while self.next_token().kind == TokenKind::Pipe {}
                    union.push(self.token.text.into());
                }

                return Ok(TypeHint::Union(union.into()));
            }
            TokenKind::Eof => return Err(ParseError::UnexpectedEnd),
            TokenKind::Pipe | TokenKind::Illegal | TokenKind::Keyword | TokenKind::Pound => {
                return Err(ParseError::IllegalToken { found: self.token })
            }
        }
    }

    fn expect(&self, kind: TokenKind) -> Result<(), ParseError<'source>> {
        if self.token.kind != kind {
            return Err(ParseError::ExpectedToken {
                expected: kind,
                found: self.token,
            });
        }

        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use crate::dotenv::typehint_parser::{Lexer, Parser};

    #[test]
    fn lexing_type_hints() {
        assert_debug_snapshot!(Lexer::new("@type string").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("# @type string").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("@type number").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("# @type number").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("@type boolean").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("@type 'qa' | 'dev' | 'prod'").collect::<Vec<_>>());
        assert_debug_snapshot!(Lexer::new("# @type 'qa' | 'dev' | 'prod'").collect::<Vec<_>>());
    }

    #[test]
    fn parse_type_hints() {
        assert_debug_snapshot!(Parser::new("@type string").parse());
        assert_debug_snapshot!(Parser::new("# @type string").parse());
        assert_debug_snapshot!(Parser::new("@type number").parse());
        assert_debug_snapshot!(Parser::new("@type boolean").parse());
        assert_debug_snapshot!(Parser::new("@type 'qa' | 'dev' | 'prod'").parse());
        assert_debug_snapshot!(Parser::new("@type 'qa' || 'dev' ||| | 'prod' | || 'test'").parse());
    }
}
