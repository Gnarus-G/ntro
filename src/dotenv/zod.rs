use colored::Colorize;
use serde_json::Value;
use std::{
    collections::BTreeMap,
    fmt::Display,
    fs::File,
    io::{BufReader, BufWriter, Write},
    ops::Deref,
    path::{Path, PathBuf},
    sync::Arc,
};

use anyhow::{anyhow, Context, Result};

use crate::command::prettify;

use super::{
    parse::{get_texts, parse_variables_with_type_hints, Variable},
    typehint_parser::TypeHint,
};

use thiserror::Error;

#[derive(Error, Debug)]
pub enum ParseError {
    #[error("\n{a}\n\nconflicts with:\n\n{b}\n")]
    ConflictingTypes { a: TypeHintAt, b: TypeHintAt },
}

#[derive(Debug)]
pub struct TypeHintAt {
    pub th: TypeHint,
    pub line: usize,
    pub meta: Metadata,
}

impl Display for TypeHintAt {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut lines = self.meta.source.lines().skip(self.line);

        let curr_line = lines
            .next()
            .expect("assumption that the type hint was parsed along with its line number, failed");
        let next_line = lines
            .next()
            .expect("assumption that the type hint was parsed along with the variable it was decorating, failed");

        writeln!(f, "{}", self.meta.path.to_string_lossy().dimmed())?;
        writeln!(
            f,
            "  {}| {}",
            self.line + 1,
            curr_line
                .replace(
                    self.th.to_string().as_str(),
                    self.th.to_string().green().to_string().as_str()
                )
                .bold()
        )?;
        write!(f, "  {}| {}", self.line + 2, next_line)?;

        Ok(())
    }
}

impl From<(&Metadata, &(TypeHint, usize))> for TypeHintAt {
    fn from(value: (&Metadata, &(TypeHint, usize))) -> Self {
        Self {
            th: value.1 .0.clone(),
            meta: value.0.clone(),
            line: value.1 .1,
        }
    }
}

#[derive(Debug, Clone)]
pub struct Metadata {
    source: Arc<str>,
    path: Arc<Path>,
}

pub fn generate_zod_schema(files: &[PathBuf]) -> Result<String> {
    let text_and_file_names = get_texts(files);

    let sources = text_and_file_names.iter().map(|(source, path)| Metadata {
        source: source.as_str().into(),
        path: path.as_path().into(),
    });

    generate_zod_schema_from_texts(sources)
}

pub fn generate_zod_schema_from_texts(sources: impl Iterator<Item = Metadata>) -> Result<String> {
    let mut map: BTreeMap<String, (Variable, Metadata)> = BTreeMap::new();

    let variables = sources.flat_map(|meta| -> Vec<(Variable, Metadata)> {
        let vars = parse_variables_with_type_hints(meta.source.deref())
            .into_iter()
            .map(|var| (var, meta.clone()))
            .collect();

        return vars;
    });

    for (var, meta) in variables {
        if let Some((v, o_meta)) = map.get(&var.key) {
            if let (Some(lt), Some(rt)) = (&v.type_hint, &var.type_hint) {
                if lt.0 != rt.0 {
                    return Err(ParseError::ConflictingTypes {
                        a: (o_meta, lt).into(),
                        b: (&meta, rt).into(),
                    })
                    .context(
                        "found some conflicting types while parsing variables with type hints",
                    );
                }
            }
        }

        map.insert(var.key.clone(), (var, meta));
    }

    let vars = map.into_values().map(|value| value.0).collect::<Vec<_>>();

    let next_public_vars = vars.iter().filter(|&v| v.is_public()).collect::<Vec<_>>();
    let other_vars = vars.iter().filter(|v| !v.is_public()).collect::<Vec<_>>();

    let to_field_schema = |var: &&Variable| -> String {
        format!(
            r#"    {}: {},"#,
            var.key,
            match &var.type_hint {
                Some(th) => match &th.0 {
                    super::typehint_parser::TypeHint::String => "z.string()".to_string(),
                    super::typehint_parser::TypeHint::Number => "z.coerce.number()".to_string(),
                    super::typehint_parser::TypeHint::Boolean => "z.coerce.boolean()".to_string(),
                    super::typehint_parser::TypeHint::Union(values) =>
                        format!("z.enum([{}])", values.join(",")),
                },
                None => "z.string()".to_string(),
            }
        )
    };

    let js_code = include_str!("module.ts");

    let js_import_line: &str = js_code
        .lines()
        .next()
        .expect("should have an import line at the top of the js implementation");

    let js_impl = js_code
        .lines()
        .skip_while(|line| !line.contains("/* --- MAIN IMPLEMENTATION BELOW --- */"))
        .skip(1)
        .collect::<Vec<_>>()
        .join("\n");

    let output = format!(
        r#"
{js_import_line}

const clientEnvSchemas = {{
{}
}}

const serverEnvSchemas = {{
    ...clientEnvSchemas,
{}
}}

{js_impl}

const processEnv = {{
{}
}}
               "#,
        next_public_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        other_vars
            .iter()
            .map(to_field_schema)
            .collect::<Vec<_>>()
            .join("\n"),
        vars.iter()
            .map(|var| format!("   {}: process.env.{},", var.key, var.key))
            .collect::<Vec<_>>()
            .join("\n"),
    );

    Ok(output)
}

pub fn add_tsconfig_path<P: AsRef<Path>>(path: P) -> Result<()> {
    let mut ts_config: Value = File::open("./tsconfig.json")
        .context("couldn't open tsconfig.json")
        .map(BufReader::new)
        .and_then(|reader| serde_json::from_reader(reader).context("failed to parse tsconfig.json"))
        .context("failed to read tsconfig.json")?;

    ts_config
        .get_mut("compilerOptions")
        .context("couldn't find compilerOptions in tsconfig.json")
        .and_then(|paths| {
            paths
                .get_mut("paths")
                .and_then(|node| node.as_object_mut())
                .map(|paths| {
                    paths.insert(
                        "$env".to_string(),
                        Value::Array(vec![Value::String(
                            path.as_ref().to_string_lossy().to_string(),
                        )]),
                    )
                })
                .ok_or(anyhow!("failed to add $env as a path on tsconfig.json"))
        })?;

    File::create("./tsconfig.json")
        .context("failed to open tsconfig.json")
        .map(BufWriter::new)
        .and_then(|mut w| {
            let new_content = serde_json::to_string::<Value>(&ts_config)?;
            let pretty = prettify(new_content.as_bytes(), ".json")?;
            w.write_all(&pretty)
                .context("failed to write to tsconfig.json")?;
            w.flush()
                .context("failed to flush updated tsconfig.json contents")
        })
        .context("failed to update tsconfig.json contents")?;

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::path::{Path, PathBuf};

    use insta::{assert_debug_snapshot, assert_display_snapshot};

    use crate::dotenv::zod::{generate_zod_schema, generate_zod_schema_from_texts};

    #[test]
    fn zod_schema_gen() {
        let output = generate_zod_schema(&[
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.test2"),
        ])
        .unwrap();
        assert_display_snapshot!(output);
    }

    #[test]
    fn zod_schema_gen_with_type_conflicts() {
        let case = |s: &str| {
            format!(
                r#"
# @type {}
KEY=
            "#,
                s
            )
        };

        fn generate(sources: &[String]) -> Result<String, anyhow::Error> {
            let sources = sources.iter().cloned().enumerate().map(|(i, source)| {
                crate::dotenv::zod::Metadata {
                    source: source.as_str().into(),
                    path: Path::new(&format!("src/dotenv/.env.test.{}", i)).into(),
                }
            });

            generate_zod_schema_from_texts(sources)
        }

        fn gen_err(sources: &[String]) {
            let output = generate(sources).unwrap_err();
            assert_debug_snapshot!(output);
        }

        gen_err(&[case("string"), case("number")]);
        gen_err(&[case("number"), case("boolean")]);
        gen_err(&[case("string"), case("boolean")]);
        gen_err(&[case("'a' | 'b'"), case("number")]);

        gen_err(&[case("string"), case("boolean"), case("number")]);

        // This is not a conflict
        generate(&[case("string"), case("string")]).unwrap();
    }
}
