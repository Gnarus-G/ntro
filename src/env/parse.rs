use chumsky::prelude::*;

#[derive(Debug, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
pub struct Variable {
    pub key: String,
}

pub fn parser() -> impl Parser<char, Vec<Variable>, Error = Simple<char>> {
    let comment = just('#')
        .then(take_until(text::newline()))
        .map(|(_, (chars, ..))| String::from_iter(chars));

    let ident = text::ident();

    let value = take_until(text::newline()).map(|(chars, ..)| chars.iter().collect::<String>());

    let line = ident.then(just('=')).then(value).map(|((key, _), _)| {
        return Variable { key };
    });

    line.padded().padded_by(comment.repeated()).repeated()
}
