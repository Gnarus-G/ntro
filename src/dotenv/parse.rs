use std::{
    fs::File,
    io::{BufReader, Read},
    path::PathBuf,
};

use anyhow::Context;

use super::typehint_parser::{ParseTyeHint, TypeHint};

type WithLineNumber<T> = (T, usize);

#[derive(Debug)]
pub struct Variable {
    pub type_hint: Option<WithLineNumber<TypeHint>>,
    pub key: String,
}

impl Variable {
    pub fn is_public(&self) -> bool {
        self.key.starts_with("NEXT_PUBLIC_")
    }
}

pub fn parse_variables(source: &str) -> Vec<&str> {
    return source
        .lines()
        .filter_map(|line| {
            if line.starts_with('#') {
                return None;
            }
            return match line.split('=').collect::<Vec<_>>()[..] {
                [ident, ..] if !ident.is_empty() => Some(ident.trim()),
                _ => None,
            };
        })
        .collect::<Vec<_>>();
}

pub fn parse_variables_with_type_hints(source: &str) -> Vec<Variable> {
    enum Token<'source> {
        LineComment(&'source str, usize),
        Ident(&'source str, usize),
    }

    let mut tokens = source
        .lines()
        .enumerate()
        .filter_map(|(l_num, line)| {
            if line.starts_with('#') {
                return Some(Token::LineComment(line, l_num));
            }
            return match line.split('=').collect::<Vec<_>>()[..] {
                [ident, ..] if !ident.is_empty() => Some(Token::Ident(ident.trim(), l_num)),
                _ => None,
            };
        })
        .peekable();

    let mut vars = Vec::new();

    loop {
        match tokens.next() {
            None => break,
            Some(token) => match (token, tokens.peek()) {
                (Token::LineComment(comment, l_num), Some(Token::Ident(ident, _))) => {
                    let var = Variable {
                        type_hint: comment.into_type_hint().map(|th| (th, l_num)),
                        key: ident.to_string(),
                    };
                    vars.push(var);
                    tokens.next();
                }
                (Token::Ident(ident, _), _) => {
                    vars.push(Variable {
                        type_hint: None,
                        key: ident.to_string(),
                    });
                }

                _ => {}
            },
        };
    }

    vars
}

pub fn get_texts(files: &[PathBuf]) -> Vec<(String, &PathBuf)> {
    files
        .iter()
        .map(|file| {
            File::open(file)
                .map(BufReader::new)
                .and_then(|mut rdr| {
                    let mut buf = String::new();
                    rdr.read_to_string(&mut buf).map(|_| buf)
                })
                .context(format!("failed read {file:?}"))
                .map(|text| (text, file))
        })
        .inspect(|result| {
            if let Err(e) = &result {
                log::error!("{e:?}");
            }
        })
        .flatten()
        .collect::<Vec<_>>()
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::assert_debug_snapshot;

    use crate::dotenv::parse::{get_texts, parse_variables_with_type_hints};

    #[test]
    fn parsing_variables_with_type_hints() {
        let sources = vec![
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.test2"),
        ];

        let output = get_texts(&sources);

        for (content, file) in output {
            let vars = parse_variables_with_type_hints(&content);
            insta::with_settings!({
                description => file.to_string_lossy()
            }, {
                assert_debug_snapshot!(vars);
            })
        }
    }
}
