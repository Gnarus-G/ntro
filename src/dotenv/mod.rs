use anyhow::{anyhow, Context, Result};
use chumsky::prelude::*;
use std::{
    collections::BTreeSet,
    fs::File,
    io::{BufReader, Read},
    path::PathBuf,
};

use self::parse::parser;

mod parse;

mod typehint_parser;
pub mod zod;

pub fn generate_typescript_types(files: &[PathBuf]) -> Result<String> {
    let parse = |text, file_name| {
        parser().parse(text).map_err(|err| {
            anyhow!(
                "failed to parse {:?}: {}",
                file_name,
                err.iter()
                    .map(|e| e.to_string())
                    .collect::<Vec<_>>()
                    .join(" ")
            )
        })
    };

    let vars = files
        .iter()
        .map(|file| {
            File::open(file)
                .map(BufReader::new)
                .and_then(|mut rdr| {
                    let mut buf = String::new();
                    rdr.read_to_string(&mut buf).map(|_| buf)
                })
                .context(format!("failed read {file:?}"))
                .and_then(|text| parse(text, file))
        })
        .filter_map(|result| {
            if let Err(e) = &result {
                eprintln!("{e:?}");
            }
            result.ok()
        })
        .flatten()
        .collect::<BTreeSet<_>>();

    let output = format!(
        r#"
declare namespace NodeJS {{
    interface ProcessEnv {{
        {}
    }}
}}
               "#,
        vars.iter()
            .map(|var| format!(
                r#"
         {}?: string"#,
                var
            ))
            .collect::<Vec<_>>()
            .join("\n")
    );

    Ok(output)
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use insta::assert_display_snapshot;

    use crate::dotenv::generate_typescript_types;

    #[test]
    fn introspect_typescript_types_gen() {
        let output = generate_typescript_types(&[
            PathBuf::from("src/dotenv/.env.test"),
            PathBuf::from("src/dotenv/.env.test2"),
        ])
        .unwrap();
        assert_display_snapshot!(output);
    }
}
