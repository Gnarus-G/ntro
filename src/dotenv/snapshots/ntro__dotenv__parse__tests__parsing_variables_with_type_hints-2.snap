---
source: src/dotenv/parse.rs
description: src/dotenv/.env.test2
expression: vars
---
[
    Variable {
        type_hint: None,
        key: "NAME5",
    },
    Variable {
        type_hint: Some(
            (
                Number,
                3,
            ),
        ),
        key: "KEY_Value",
    },
    Variable {
        type_hint: Some(
            (
                Union(
                    [
                        "'city'",
                        "'townhall'",
                    ],
                ),
                6,
            ),
        ),
        key: "keys2Da",
    },
    Variable {
        type_hint: Some(
            (
                Union(
                    [
                        "'val'",
                        "'value'",
                    ],
                ),
                9,
            ),
        ),
        key: "NAME",
    },
]
